File: backend/app/utils/exceptions.py
from fastapi import HTTPException, status

class CodingPlatformException(HTTPException):
    def __init__(self, status_code: int, code: str, message: str):
        super().__init__(status_code=status_code, detail={"code": code, "message": message})

class InvalidLanguageException(CodingPlatformException):
    def __init__(self, language: str):
        super().__init__(
            status_code=status.HTTP_400_BAD_REQUEST,
            code="INVALID_LANGUAGE",
            message=f"Language '{language}' is not supported. Supported: python, javascript, sql"
        )

class QuestionNotFoundException(CodingPlatformException):
    def __init__(self, question_id: int):
        super().__init__(
            status_code=status.HTTP_404_NOT_FOUND,
            code="QUESTION_NOT_FOUND",
            message=f"Question with ID {question_id} not found"
        )

class TestCaseNotFoundException(CodingPlatformException):
    def __init__(self, testcase_id: int):
        super().__init__(
            status_code=status.HTTP_404_NOT_FOUND,
            code="TESTCASE_NOT_FOUND",
            message=f"TestCase with ID {testcase_id} not found"
        )

class SubmissionNotFoundException(CodingPlatformException):
    def __init__(self, submission_id: int):
        super().__init__(
            status_code=status.HTTP_404_NOT_FOUND,
            code="SUBMISSION_NOT_FOUND",
            message=f"Submission with ID {submission_id} not found"
        )

class Judge0UnavailableException(CodingPlatformException):
    def __init__(self, detail: str = "Judge0 execution service is unavailable"):
        super().__init__(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            code="JUDGE0_UNAVAILABLE",
            message=detail
        )

File: backend/app/common/utils/exceptions.js
class CodingPlatformException extends Error {
  constructor(statusCode, code, message) {
    super(message);
    this.statusCode = statusCode;
    this.code = code;
  }
}

class InvalidLanguageException extends CodingPlatformException {
  constructor(language) {
    super(400, 'INVALID_LANGUAGE', `Language '${language}' is not supported. Supported: python, javascript, sql`);
  }
}

class QuestionNotFoundException extends CodingPlatformException {
  constructor(questionId) {
    super(404, 'QUESTION_NOT_FOUND', `Question with ID ${questionId} not found`);
  }
}

class TestCaseNotFoundException extends CodingPlatformException {
  constructor(testcaseId) {
    super(404, 'TESTCASE_NOT_FOUND', `TestCase with ID ${testcaseId} not found`);
  }
}

class SubmissionNotFoundException extends CodingPlatformException {
  constructor(submissionId) {
    super(404, 'SUBMISSION_NOT_FOUND', `Submission with ID ${submissionId} not found`);
  }
}

class Judge0UnavailableException extends CodingPlatformException {
  constructor(detail = 'Judge0 execution service is unavailable') {
    super(503, 'JUDGE0_UNAVAILABLE', detail);
  }
}

// New module exceptions
class QuizNotFoundException extends CodingPlatformException {
  constructor(quizId) {
    super(404, 'QUIZ_NOT_FOUND', `Quiz with ID ${quizId} not found`);
  }
}

class AssignmentNotFoundException extends CodingPlatformException {
  constructor(assignmentId) {
    super(404, 'ASSIGNMENT_NOT_FOUND', `Assignment with ID ${assignmentId} not found`);
  }
}

class BugfixChallengeNotFoundException extends CodingPlatformException {
  constructor(challengeId) {
    super(404, 'BUGFIX_CHALLENGE_NOT_FOUND', `Bug fix challenge with ID ${challengeId} not found`);
  }
}

class DeadlinePassedException extends CodingPlatformException {
  constructor(assignmentId) {
    super(400, 'DEADLINE_PASSED', `Deadline for assignment ${assignmentId} has passed`);
  }
}

class AttemptLimitExceededException extends CodingPlatformException {
  constructor(maxAttempts) {
    super(400, 'ATTEMPT_LIMIT_EXCEEDED', `Maximum attempts (${maxAttempts}) exceeded`);
  }
}

module.exports = {
  CodingPlatformException,
  InvalidLanguageException,
  QuestionNotFoundException,
  TestCaseNotFoundException,
  SubmissionNotFoundException,
  Judge0UnavailableException,
  QuizNotFoundException,
  AssignmentNotFoundException,
  BugfixChallengeNotFoundException,
  DeadlinePassedException,
  AttemptLimitExceededException
};
